pkg/controllers/vpc/vpc_controller.go

```diff
@@ -0,0 +1,214 @@
+/* Copyright © 2023 VMware, Inc. All Rights Reserved.
+   SPDX-License-Identifier: Apache-2.0 */
+
+package vpc
+
+import (
+	"context"
+	"errors"
+	"os"
+	"runtime"
+	"strings"
+	"time"
+
+	apimachineryruntime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/vmware-tanzu/nsx-operator/pkg/apis/v1alpha1"
+	"github.com/vmware-tanzu/nsx-operator/pkg/controllers/common"
+	"github.com/vmware-tanzu/nsx-operator/pkg/logger"
+	"github.com/vmware-tanzu/nsx-operator/pkg/metrics"
+	_ "github.com/vmware-tanzu/nsx-operator/pkg/nsx/ratelimiter"
+	commonservice "github.com/vmware-tanzu/nsx-operator/pkg/nsx/services/common"
+	"github.com/vmware-tanzu/nsx-operator/pkg/nsx/services/vpc"
+)
+
+var (
+	log                     = logger.Log
+	ResultNormal            = common.ResultNormal
+	ResultRequeue           = common.ResultRequeue
+	ResultRequeueAfter5mins = common.ResultRequeueAfter5mins
+	MetricResType           = common.MetricResTypeVPC
+)
+
+// VPCReconciler VPCReconcile reconciles a VPC object
+type VPCReconciler struct {
+	Client  client.Client
+	Scheme  *apimachineryruntime.Scheme
+	Service *vpc.VPCService
+}
+
+func (r *VPCReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	obj := &v1alpha1.VPC{}
+	log.Info("reconciling VPC CR", "vpc", req.NamespacedName)
+	metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerSyncTotal, common.MetricResTypeVPC)
+
+	if err := r.Client.Get(ctx, req.NamespacedName, obj); err != nil {
+		log.Error(err, "unable to fetch vpc CR", "req", req.NamespacedName)
+		return common.ResultNormal, client.IgnoreNotFound(err)
+	}
+
+	// Since VPC service can only be activated from NSX 4.1.0 onwards,
+	// So need to check NSX version before starting VPC reconciler
+	// TODO: just reuse same function from static route.
+	/*
+		if !r.Service.NSXClient.NSXCheckVersionForStaticRoute() {
+			err := errors.New("NSX version check failed, VPC feature is not supported")
+			updateFail(r, &ctx, obj, &err)
+			// if NSX version check fails, it will be put back to reconcile queue and be reconciled after 5 minutes
+			return ResultRequeueAfter5mins, nil
+		}
+	*/
```

Comment:
Remove the commented code, we need to check NSX version in the beginning when start VPC related controllers